src/fairseq2/models/transformer/experts.py

```diff
@@ -0,0 +1,328 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, final
+
+import torch
+from torch import Tensor
+from torch.nn import Module, SiLU
+from torch.nn.parameter import Parameter
+
+from fairseq2.data_type import DataType
+from fairseq2.device import META_DEVICE, Device
+from fairseq2.gang import Gang
+from fairseq2.nn.utils.module import to_empty
+from fairseq2.ops.tensor_parallel import reduce, reduce_on_backward
+
+
+class ExpertNetwork(Module, ABC):
+    def __init__(
+        self,
+        num_experts: int,
+        model_dim: int,
+        inner_dim: int,
+    ) -> None:
+        super().__init__()
+        self.num_experts = num_experts
+        self.model_dim = model_dim
+        self.inner_dim = inner_dim
+
+    @abstractmethod
+    def forward(self, x: Tensor) -> Tensor:
+        """
+        :param x: Input tensor of shape (num_local_experts, tokens_per_expert, dim).
+        :returns: Output tensor of shape (num_local_experts, tokens_per_expert, dim).
+        """
+
+    @property
+    def num_local_experts(self) -> int:
+        """
+        The number of experts local to the current process.
+        Can be overriden in expert-parallel implementations.
+        """
+        return self.num_experts
+
+    def extra_repr(self) -> str:
+        """:meta private:"""
+        local_s = (
+            f"num_local_experts={self.num_local_experts}, "
+            if self.num_local_experts != self.num_experts
+            else ""
+        )
+        return f"num_experts={self.num_experts}, {local_s}model_dim={self.model_dim}, inner_dim={self.inner_dim}"
+
+    if TYPE_CHECKING:
+        __call__ = forward
+
+
+class _GLUExpertNetwork:
+    """A mixin for GLU-based experts.
+    Their layers are implemented as 3-D tensors used in BMM.
+    """
+
+    def _create_layers(
```

Comment:
Nit: this can be a standalone function like `create_expert_glu_layers` instead of a mixin.